System test: readerCheckpointTest failed

**Problem description**
We have observed a test failure in a run of the system tests in which the last event was not read after the second time we reset to a checkpoint in the test. Specifically, the readers did not read event 299, although it had been read during the previous reset to a checkpoint:

```
2017-12-02 21:48:28,399 83521 [main] ERROR i.p.t.s.f.TestFrameworkException - TestFramework Exception. Type: InternalError, Details: Exception while running test method: readerCheckpointTest
java.lang.AssertionError: Check for last event: 298, 300
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at io.pravega.test.system.ReaderCheckpointTest.verifyEvents(ReaderCheckpointTest.java:235)
        at io.pravega.test.system.ReaderCheckpointTest.readEventsAndVerify(ReaderCheckpointTest.java:217)
        at io.pravega.test.system.ReaderCheckpointTest.readerCheckpointTest(ReaderCheckpointTest.java:177)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
```

Attaching the logs for the test case.

[readercheckpointtest.txt](https://github.com/pravega/pravega/files/1524651/readercheckpointtest.txt)

**Problem location**
System tests, resetting to a checkpoint.

**Suggestions for an improvement**
Track and fix the bug.

--- a/test/system/src/test/java/io/pravega/test/system/ReaderCheckpointTest.java
+++ b/test/system/src/test/java/io/pravega/test/system/ReaderCheckpointTest.java
@@ -262,7 +262,8 @@ public class ReaderCheckpointTest {
                     log.error("Exception while reading event using readerId: {}", readerId, e);
                     fail("Reinitialization Exception is not expected");
                 }
-            } while (event.getEvent() != null);
+            } while (event.isCheckpoint() || event.getEvent() != null);
+            //stop reading if event read(non-checkpoint) is null.
             log.info("No more events from {}/{} for readerId: {}", SCOPE, STREAM, readerId);
         } //reader.close() will automatically invoke ReaderGroup#readerOffline(String, Position)
         return events;

--- a/test/system/src/test/java/io/pravega/test/system/ReaderCheckpointTest.java
+++ b/test/system/src/test/java/io/pravega/test/system/ReaderCheckpointTest.java
@@ -258,6 +258,9 @@ public class ReaderCheckpointTest {
                         log.info("Read event {}", event.getEvent());
                         events.add(event);
                     }
+                    if (event.isCheckpoint()) {
+                        log.info("Read a check point event, checkpointName: {}", event.getCheckpointName());
+                    }
                 } catch (ReinitializationRequiredException e) {
                     log.error("Exception while reading event using readerId: {}", readerId, e);
                     fail("Reinitialization Exception is not expected");

